File: Actividad 4/src/java/Benchmark.java
/*
 * Benchmark.java
 * --------------------------------------------------------------------------
 * Driver de ejecución en Java para el benchmark de lexer/parser generado
 * por ANTLR4 a partir de grammar/YamlDocker.g4.
 *
 * Uso:
 *     java -cp .:antlr-4.13.1-complete.jar Benchmark <ruta_al_archivo.yml>
 *
 * Salida:
 *     Imprime EXCLUSIVAMENTE en stdout el tiempo transcurrido en
 *     milisegundos (double), sin texto adicional, para que
 *     scripts/run_benchmark.py pueda capturarlo directamente vía
 *     subprocess.
 *
 * Requisitos previos:
 *     - Haber generado el lexer/parser Java con:
 *         antlr4 -Dlanguage=Java -visitor -o generated/java grammar/YamlDocker.g4
 *       (ver Makefile, target `build`)
 *     - Haber compilado tanto el código generado como este driver contra
 *       antlr-4.13.1-complete.jar (runtime + herramienta en un solo jar).
 *
 * Nota sobre la medición:
 *     Se usa System.nanoTime() por ser el reloj monotónico de mayor
 *     resolución disponible en la JVM (no afectado por ajustes del reloj
 *     de pared). El resultado se convierte a milisegundos como double para
 *     mantener precisión sub-milisegundo, igual que time.perf_counter() en
 *     Python y std::chrono::high_resolution_clock en C++.
 */

import org.antlr.v4.runtime.CharStream;
import org.antlr.v4.runtime.CharStreams;
import org.antlr.v4.runtime.CommonTokenStream;
import org.antlr.v4.runtime.tree.ParseTree;

import java.io.PrintStream;
import java.nio.charset.StandardCharsets;

public class Benchmark {

    public static void main(String[] args) {
        if (args.length != 1) {
            System.err.println("Uso: java Benchmark <archivo.yml>");
            System.exit(1);
        }

        String filepath = args[0];

        try {
            double elapsedMs = parseFile(filepath);
            // ÚNICA salida por stdout: el tiempo en milisegundos
            PrintStream out = new PrintStream(System.out, true, "UTF-8");
            out.printf("%.4f%n", elapsedMs);
        } catch (Exception e) {
            System.err.println("ERROR al parsear '" + filepath + "': " + e.getMessage());
            e.printStackTrace(System.err);
            System.exit(1);
        }
    }

    /**
     * Realiza el lexing + parsing completo del archivo dado y devuelve el
     * tiempo transcurrido en milisegundos.
     */
    private static double parseFile(String filepath) throws Exception {
        long start = System.nanoTime();

        CharStream input = CharStreams.fromFileName(filepath, StandardCharsets.UTF_8);
        YamlDockerLexer lexer = new YamlDockerLexer(input);
        lexer.removeErrorListeners(); // evita ruido de errores en stdout/stderr

        CommonTokenStream tokens = new CommonTokenStream(lexer);
        YamlDockerParser parser = new YamlDockerParser(tokens);
        parser.removeErrorListeners();

        ParseTree tree = parser.document(); // regla inicial de la gramática
        if (tree == null) {
            throw new IllegalStateException("El parser no devolvió un árbol válido");
        }

        long end = System.nanoTime();
        return (end - start) / 1_000_000.0; // nanosegundos -> milisegundos
    }
}

File: Actividad 4/src/cpp/benchmark.cpp
/*
 * benchmark.cpp
 * --------------------------------------------------------------------------
 * Driver de ejecución en C++ para el benchmark de lexer/parser generado
 * por ANTLR4 a partir de grammar/YamlDocker.g4.
 *
 * Uso:
 *     ./benchmark <ruta_al_archivo.yml>
 *
 * Salida:
 *     Imprime EXCLUSIVAMENTE en stdout el tiempo transcurrido en
 *     milisegundos (double), sin texto adicional, para que
 *     scripts/run_benchmark.py pueda capturarlo directamente vía
 *     subprocess.
 *
 * Requisitos previos:
 *     - Haber generado el lexer/parser C++ con:
 *         antlr4 -Dlanguage=Cpp -visitor -o generated/cpp grammar/YamlDocker.g4
 *       (ver Makefile, target `build`)
 *     - Tener instalado antlr4-cpp-runtime (headers + libantlr4-runtime.so/.a)
 *     - Compilar este archivo junto con el código generado, enlazando contra
 *       antlr4-runtime (ver Makefile, target `build`).
 *
 * Nota sobre la medición:
 *     Se usa std::chrono::high_resolution_clock, el reloj de mayor
 *     resolución disponible en <chrono>, equivalente en propósito a
 *     time.perf_counter() (Python) y System.nanoTime() (Java).
 */

#include <iostream>
#include <chrono>
#include <string>
#include <cstdio>

#include "antlr4-runtime.h"
#include "YamlDockerLexer.h"
#include "YamlDockerParser.h"

using namespace antlr4;

// Listener de errores "silencioso": evita que ANTLR imprima diagnósticos en
// stderr/stdout durante la medición, igual que removeErrorListeners() en
// las versiones Python y Java del driver.
class SilentErrorListener : public BaseErrorListener {
public:
    void syntaxError(Recognizer * /*recognizer*/, Token * /*offendingSymbol*/,
                      size_t /*line*/, size_t /*charPositionInLine*/,
                      const std::string & /*msg*/, std::exception_ptr /*e*/) override {
        // Intencionalmente vacío: se descartan los errores de sintaxis
        // para que no contaminen stdout durante la medición de tiempos.
    }
};

static double parseFile(const std::string &filepath) {
    auto start = std::chrono::high_resolution_clock::now();

    std::ifstream stream;
    stream.open(filepath);
    if (!stream.is_open()) {
        throw std::runtime_error("No se pudo abrir el archivo: " + filepath);
    }

    ANTLRInputStream input(stream);
    YamlDockerLexer lexer(&input);

    SilentErrorListener silentListener;
    lexer.removeErrorListeners();
    lexer.addErrorListener(&silentListener);

    CommonTokenStream tokens(&lexer);
    YamlDockerParser parser(&tokens);
    parser.removeErrorListeners();
    parser.addErrorListener(&silentListener);

    tree::ParseTree *tree = parser.document(); // regla inicial de la gramática
    if (tree == nullptr) {
        throw std::runtime_error("El parser no devolvio un arbol valido");
    }

    auto end = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double, std::milli> elapsed = end - start;
    return elapsed.count();
}

int main(int argc, char *argv[]) {
    if (argc != 2) {
        std::cerr << "Uso: " << argv[0] << " <archivo.yml>" << std::endl;
        return 1;
    }

    std::string filepath = argv[1];

    try {
        double elapsedMs = parseFile(filepath);
        // UNICA salida por stdout: el tiempo en milisegundos
        std::printf("%.4f\n", elapsedMs);
    } catch (const std::exception &e) {
        std::cerr << "ERROR al parsear '" << filepath << "': " << e.what() << std::endl;
        return 1;
    }

    return 0;
}
